README.md
# UTIL-DISK-Distributed_File_Repair_Utilities
Set of tools is designed to help people repair a corrupt file with the help of someone who got the original file.

LICENSE
===============
* GNU GPL V2

COMPILE TOOLS
===============
* 
 
AUTHORS
===============
* Loki TroubleMind

LINKS
===============
* 


CHANGELOG
CHANGELOG
--------------

0.2.0b:
- used MD5 instead of CRC32
- hashes are saved binary
* WARNING!*
* completely incompatible to 0.1.5b!!! *

0.1.5b:
- first public release


$loki 2006-01-21


makehashes.c
	/*****************************************************************************
 *
 *  Distributed File Repair Utilities v0.1.5 beta
 *  makehashes.c - produces hashlist of a (probably) corrupt infile
 *
 *  Copyright (C) 2005 Loki TroubleMind <loki1985@users.sourceforge.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 *
 * $loki 2005-04-05
 *
 ****************************************************************************/

#include <stdio.h>
#include <string.h>
#include "md5.h"

#define CHUNKSIZE 1024 * 32

void processnextchunk(void);

char infilename[255] = "";
char outfilename[255] = "hashlist";

FILE *ifp;
FILE *ofp;

long ifoffset = 0;
long ifsize = 0;
long curchunklen = 0;
int eof_reached = 0;

md5_state_t state;
md5_byte_t digest[16];

unsigned char chunkdata[CHUNKSIZE] = {0};

int main(int argc, char *argv[]) {

    if (argc < 2) {printf("usage: %s infile\n", argv[0]); return -1;}

    strcpy(infilename, argv[1]);

    if(NULL != (ofp = fopen(outfilename, "r"))) {printf("ERROR: '%s' already exists\n", outfilename); return -1;}

    if(NULL == (ifp = fopen(infilename, "rb"))) {printf("ERROR: cannot open '%s'\n", infilename); return -1;}
    if(NULL == (ofp = fopen(outfilename, "ab"))) {printf("ERROR: cannot write to '%s'\n", outfilename); return -1;}

    fseek(ifp, 0, SEEK_END);
    ifsize = ftell(ifp);
    fseek(ifp, 0, SEEK_SET);

    while(eof_reached != 1) {

      processnextchunk();

      printf("\rdone: %d / %d", ifoffset -1, ifsize);
    }

    fclose (ifp);
    fclose (ofp);

    return 0;
}


void processnextchunk() {

    int bytetemp = 0;
    int bytecount = 0;
    int i;

    curchunklen = 0;

    while((bytecount < CHUNKSIZE) && (eof_reached != 1)) {

      if((bytetemp = fgetc(ifp)) == EOF) {eof_reached = 1;}
      else {
        chunkdata[bytecount] = (unsigned char)bytetemp;
        bytecount++;
        curchunklen++;
      }

      ifoffset++;
    }

	md5_init(&state);
	md5_append(&state, (const md5_byte_t *)chunkdata, curchunklen);
	md5_finish(&state, digest);

    for(i=0;i<16;i++) {fputc(digest[i], ofp);}

    return;
}


insertchunks.c
/*****************************************************************************
 *
 *  Distributed File Repair Utilities v0.1.5 beta
 *  insertchunks.c - inserts .fix-chunks a copy of the damaged file...
 *
 *  Copyright (C) 2005 Loki TroubleMind <loki1985@users.sourceforge.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 *
 * $loki 2005-04-05
 *
 ****************************************************************************/

#include <stdio.h>
#include <string.h>

#define CHUNKSIZE 1024 * 32

char infilename[255] = "";
char outfilename[255] = "";
char fixfilename[255] = "";

FILE *ifp;
FILE *ofp;
FILE *ffp;

long currentoffset = 0;
long currentchunkoffset = 0;
unsigned int currentbyte = 0;
long ifsize = 0;
int chunknum = 0;
int eof_reached = 0;

unsigned char chunkdata[CHUNKSIZE] = {0};

int main(int argc, char *argv[]) {

    int bytetemp = 0;
    int bytecount = 0;
    long outbytecount;

    if (argc < 2) {printf("usage: %s brokenfile\n", argv[0]); return -1;}

    strcpy(infilename, argv[1]);
    sprintf(outfilename, "%s_fixed", infilename);

    if(NULL != (ofp = fopen(outfilename, "r"))) {printf("ERROR: '%s' already exists\n", outfilename); return -1;}

    if(NULL == (ifp = fopen(infilename, "rb"))) {printf("ERROR: cannot open '%s'\n", infilename); return -1;}
    if(NULL == (ofp = fopen(outfilename, "ab"))) {printf("ERROR: cannot open '%s'\n", infilename); return -1;}

    fseek(ifp, 0, SEEK_END);
    ifsize = ftell(ifp);
    fseek(ifp, 0, SEEK_SET);

    while (currentoffset < ifsize) {

      bytetemp = 0;
      bytecount = 0;

      sprintf(fixfilename, "%08d.fix", chunknum);

      while(bytecount <= (CHUNKSIZE - 1) && eof_reached != 1) {

        if((bytetemp = fgetc(ifp)) == EOF) {eof_reached = 1;}
        else {
          chunkdata[bytecount] = (unsigned short)bytetemp;
          bytecount++;
        }

        currentoffset++;
      }

      if(NULL != (ffp = fopen(fixfilename, "rb"))) {

        bytetemp = 0;
        bytecount = 0;

        eof_reached = 0;

        while(bytecount <= (CHUNKSIZE - 1) && eof_reached != 1) {

          if((bytetemp = fgetc(ffp)) == EOF) {eof_reached = 1;}
          else {
            chunkdata[bytecount] = (unsigned short)bytetemp;
            bytecount++;
          }
        }

        fclose(ffp);
      }

      outbytecount = 0;

      while(outbytecount < CHUNKSIZE && outbytecount < bytecount) {
        fputc(chunkdata[outbytecount], ofp);
        outbytecount++;
      }

      printf("\rdone: %d / %d", currentoffset - 1, ifsize);

      chunknum++;
    }

    fclose (ifp);
    fclose (ofp);

    return 0;
}



makechunks.c
/*****************************************************************************
 *
 *  Distributed File Repair Utilities v0.1.5 beta
 *  makechunks.c - compares datafile by hashlist & extracts differing chunks
 *
 *  Copyright (C) 2005 Loki TroubleMind <loki1985@users.sourceforge.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 *
 * $loki 2005-04-05
 *
 ****************************************************************************/

#include <stdio.h>
#include <string.h>
#include "md5.h"

#define CHUNKSIZE 1024 * 32

void processnextchunk(void);

char infilename[255] = "";
char checkfilename[255] = "hashlist";
char outfilename[255] = "";

FILE *ifp;
FILE *cfp;

md5_state_t state;
md5_byte_t digest[16];

long ifoffset = 0;
long cfoffset = 0;
long ifsize = 0;
long curchunklen = 0;
int eof_reached = 0;
int chunknum = 0;
int brokenchunks = 0;

unsigned char chunkdata[CHUNKSIZE] = {0};

int main(int argc, char *argv[]) {

    if (argc < 2) {printf("usage: %s originalfile\n", argv[0]); return -1;}

    strcpy(infilename, argv[1]);

    if(!(ifp = fopen(infilename, "rb"))) {printf("ERROR: cannot open '%s'\n", infilename); return -1;}
    if(!(cfp = fopen(checkfilename, "rb"))) {printf("ERROR: cannot open '%s'\n", checkfilename); return -1;}

    fseek(ifp, 0, SEEK_END);
    ifsize = ftell(ifp);
    fseek(ifp, 0, SEEK_SET);

    while(eof_reached != 1) {

      processnextchunk();

      printf("\rdone: %d / %d", ifoffset, ifsize);

      chunknum++;
    }

    printf("\ntotal chunks: %d\nbroken chunks: %d\n\n", chunknum, brokenchunks);

    fclose (ifp);
    fclose (cfp);

    return 0;
}


void processnextchunk() {

    FILE *ofp;

    int bytetemp = 0;
    int bytecount = 0;
    int different;
    int i;
    int outbytecount = 0;

    unsigned char loadedhash[16];

    curchunklen = 0;

    while((bytecount <= CHUNKSIZE - 1) && (eof_reached != 1)) {

      if((bytetemp = fgetc(ifp)) == EOF) {eof_reached = 1;}
      else {
        chunkdata[bytecount] = (unsigned short)bytetemp;
        bytecount++;
        curchunklen++;
      }

      ifoffset++;
    }

    md5_init(&state);
    md5_append(&state, (const md5_byte_t *)chunkdata, curchunklen);
    md5_finish(&state, digest);

	for(i=0;i<16;i++) {
      loadedhash[i] = (unsigned char)fgetc(cfp);
    }

    different = 0;

	for(i=0;i<16;i++) {
	  if(digest[i] != loadedhash[i]) {different = 1;}
    }

    if(different) {

      brokenchunks++;

      sprintf(outfilename, "%08d.fix", chunknum);

      if(!(ofp = fopen(outfilename, "ab"))) {printf("ERROR: cannot write to '%s'\n", outfilename); return;}

      outbytecount = 0;

      while(outbytecount < curchunklen) {
        fputc(chunkdata[outbytecount], ofp);
        outbytecount++;
      }

      fclose(ofp);
    }

    return;
}
